File: GMUtilsAndroid/GMUtils/GMUtils/src/main/java/gmutils/firebase/database/FirebaseDatabaseOp.java
package gmutils.firebase.database;

import android.text.TextUtils;
import android.util.Log;
import android.util.Pair;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.GenericTypeIndicator;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import gmutils.StringSet;
import gmutils.backgroundWorkers.BackgroundTaskAbs;
import gmutils.firebase.FirebaseUtils;
import gmutils.firebase.Response;
import gmutils.listeners.ActionCallback;
import gmutils.listeners.ResultCallback;

public class FirebaseDatabaseOp extends IFirebaseDatabaseOp {
    public final DatabaseReference databaseReference;

    public FirebaseDatabaseOp() {
        this("");
    }

    public FirebaseDatabaseOp(String rootNodeName) {
        try {
            Class.forName("com.google.firebase.database.FirebaseDatabase");
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            throw new IllegalStateException("add this line to gradle script file:\n" +
                    "//https://firebase.google.com/docs/database/android/start\n" +
                    "implementation 'com.google.firebase:firebase-database:19.6.0'");
        }

        if (TextUtils.isEmpty(rootNodeName)) {
            databaseReference = FirebaseDatabase.getInstance().getReference();
        } else {
            databaseReference = FirebaseDatabase.getInstance().getReference(rootNodeName);
        }
    }

    //----------------------------------------------------------------------------

    private DatabaseReference _getReferenceOfNode(String subNodePath) {
        if (TextUtils.isEmpty(subNodePath)) return databaseReference;

        try {
            subNodePath = FirebaseUtils.refinePathFragmentNames(subNodePath);
        } catch (Exception e) {
            throw new RuntimeException(
                    "Exception at _getReferenceOfNode: ",
                    e
            );
        }

        var ref = databaseReference;
        return ref.child(subNodePath);
    }

    //----------------------------------------------------------------------------

    @Override
    public <T> void saveData(
            T data,
            @Nullable String subNodePath,
            ResultCallback<Response<Boolean>> callback
    ) {
        DatabaseReference ref = _getReferenceOfNode(subNodePath);
        saveData(ref, data, callback);
    }

    public <T> void saveData(
            DatabaseReference ref,
            T data,
            ResultCallback<Response<Boolean>> callback
    ) {
        ref.setValue(data)
                .addOnSuccessListener(unused -> {
                    if (callback != null) callback.invoke(Response.success(true));
                })
                .addOnFailureListener(e -> {
                    if (callback != null)
                        callback.invoke(Response.failed(new StringSet(e.getMessage())));
                });
    }

    @Override
    public <T> void saveMultipleData(
            Map<String, T> nodesAndData,
            ResultCallback<Response<Boolean>> callback
    ) {
        String[] keys = nodesAndData.keySet().toArray(new String[]{});
        if (keys.length > 0) {
            int i = 0;
            saveData(nodesAndData, keys, i, callback);
        }

//        DatabaseReference ref = _getReferenceOfNode(null);
//        saveMultipleData(ref, nodesAndData, callback);
    }

    private <T> void saveData(
            Map<String, T> nodesAndData,
            String[] keys,
            int kidx,
            ResultCallback<Response<Boolean>> callback
    ) {
        saveData(nodesAndData.get(keys[kidx]), keys[kidx], result -> {
            if (result.data != Boolean.TRUE) {
                callback.invoke(result);
            } else {
                int idx = kidx + 1;
                if (idx < keys.length) {
                    saveData(nodesAndData, keys, idx, callback);
                } else {
                    callback.invoke(result);
                }
            }
        });
    }

    //----------------------------------------------------------------------------

    @Override
    public <T> void retrieveAll(
            FBFilterOption filterOption,
            Class<T> dataType,
            ActionCallback<Object, List<T>> customConverter,
            ResultCallback<Response<List<T>>> callback
    ) {
        var ref = databaseReference;

        ValueEventListener eventListener = new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                if (snapshot.exists()) {
                    backgroundTask().execute(() -> {
                        List<T> list = new ArrayList<>();
                        Object value = snapshot.getValue();

                        if (value != null) {
                            if (customConverter == null) {
                                try {
                                    if (value instanceof List) {
                                        GenericTypeIndicator<List<T>> type = new GenericTypeIndicator<List<T>>() {
                                        };
                                        List<T> values = snapshot.getValue(type);
                                        list.addAll(values);
                                    }
                                    //
                                    else if (value instanceof Map) {
                                        Map valueAsMap = (Map) value;
                                        try {
                                            for (Object key : valueAsMap.keySet()) {
                                                Object o = valueAsMap.get(key);
                                                String json = new Gson().toJson(o);
                                                T t = new Gson().fromJson(json, dataType);
                                                list.add(t);
                                            }
                                        } catch (Exception e) {
                                            String json = new Gson().toJson(valueAsMap);
                                            T t = new Gson().fromJson(json, dataType);
                                            list.add(t);
                                        }
                                    }
                                    //
                                    else {
                                        T value2 = snapshot.getValue(dataType);
                                        list.add(value2);
                                    }
                                } catch (Exception e) {
                                    throw new RuntimeException(
                                            "Exception at retrieveAll: " + e.getMessage() + "\n--------\n" + value,
                                            e
                                    );
                                }
                            }
                            //
                            else {
                                var values = customConverter.invoke(value);
                                list.addAll(values);
                            }
                        }

                        return list;
                    }, list -> {
                        callback.invoke(Response.success(list));
                    });
                }
                //
                else {
                    callback.invoke(
                            Response.failed(
                                    new StringSet("No data", "لا توجد بيانات")
                            )
                    );
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                callback.invoke(Response.failed(new StringSet(error.getMessage())));
            }
        };

        if (filterOption == null) {
            ref.addListenerForSingleValueEvent(eventListener);
        } else {
            Query query = buildQuery(ref, filterOption);
            query.addListenerForSingleValueEvent(eventListener);
        }
    }

    protected BackgroundTaskAbs backgroundTask() {
        return BackgroundTaskAbs.getInstance();
    }

    //----------------------------------------------------------------------------

    @Override
    public <T> void retrieveSingle(
            String subNodePath,
            Class<T> dataType,
            ActionCallback<Object, Pair<T, StringSet>> customConverter,
            ResultCallback<Response<T>> callback
    ) {
        var ref = _getReferenceOfNode(subNodePath);
        retrieveSingle(ref, dataType, customConverter, callback);
    }

    public <T> void retrieveSingle(
            DatabaseReference ref,
            Class<T> dataType,
            ActionCallback<Object, Pair<T, StringSet>> customConverter,
            ResultCallback<Response<T>> callback
    ) {
        ref.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                handleReceivedDataSnapshot(snapshot, dataType, customConverter, callback);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                callback.invoke(Response.failed(new StringSet(error.getMessage())));
            }
        });
    }

    private <T2> void handleReceivedDataSnapshot(
            DataSnapshot snapshot,
            Class<T2> dataType,
            ActionCallback<Object, Pair<T2, StringSet>> customConverter,
            ResultCallback<Response<T2>> callback
    ) {
        if (snapshot.exists()) {
            backgroundTask().execute(() -> {
                Pair<T2, StringSet> data = null;

                if (customConverter == null) {
                    if (snapshot.getValue() != null) {
                        T2 obj = null;

                        try {
                            obj = snapshot.getValue(dataType);
                        } catch (Exception e) {
                            throw new RuntimeException(
                                    "Exception at handleReceivedDataSnapshot: " + e.getMessage() + "\nValue: " + snapshot.getValue(),
                                    e
                            );
                        }

                        if (obj != null) {
                            data = new Pair<>(obj, null);
                        } else {
                            data = new Pair<>(
                                    null,
                                    new StringSet(
                                            "Invalid data type (" + dataType + ") ... exist object => " + snapshot.getValue()
                                    )
                            );
                        }
                    } else {
                        data = new Pair<>(null, new StringSet("No data..", "لا توجد بيانات.."));
                    }
                } else {
                    data = customConverter.invoke(snapshot.getValue());
                }

                return data;
            }, data -> {
                if (data.first != null) {
                    callback.invoke(Response.success(data.first));
                } else {
                    callback.invoke(Response.failed(data.second));
                }
            });
        }

        //
        else {
            callback.invoke(
                    Response.failed(
                            new StringSet("No data", "لا توجد بيانات")
                    )
            );
        }
    }

    //----------------------------------------------------------------------------

    @Override
    public <T> void listenToChanges(
            String subNodePath,
            Class<T> dataType,
            FBFilterOption filterOption,
            ResultCallback<Updates<T>> onChange,
            ResultCallback<String> onError
    ) {
        listenToChanges(
                _getReferenceOfNode(subNodePath),
                dataType,
                filterOption,
                onChange,
                onError
        );
    }

    public <T> void listenToChanges(
            @NotNull DatabaseReference ref,
            Class<T> dataType,
            FBFilterOption filterOption,
            ResultCallback<Updates<T>> onChange,
            ResultCallback<String> onError
    ) {
        _listenTo(
                ref,
                true,
                dataType,
                filterOption,
                onChange,
                onError
        );
    }

    @Override
    public <N> void listenToChangesSpecific(
            String subNodePath,
            Class<N> dataType,
            FBFilterOption filterOption,
            ResultCallback<Updates<N>> onChange,
            ResultCallback<String> onError
    ) {
        _listenTo(
                _getReferenceOfNode(subNodePath),
                true,
                dataType,
                filterOption,
                onChange,
                onError
        );
    }

    @Override
    public <T> void listenToAdding(
            String subNodePath,
            Class<T> dataType,
            FBFilterOption filterOption,
            ResultCallback<Updates<T>> onAdd,
            ResultCallback<String> onError
    ) {
        listenToAdding(
                _getReferenceOfNode(subNodePath),
                dataType,
                filterOption,
                onAdd,
                onError
        );
    }

    public <T> void listenToAdding(
            DatabaseReference ref,
            Class<T> dataType,
            FBFilterOption filterOption,
            ResultCallback<Updates<T>> onAdd,
            ResultCallback<String> onError
    ) {
        _listenTo(
                ref,
                false,
                dataType,
                filterOption,
                onAdd,
                onError
        );
    }

    @Override
    public <N> void listenToAddingSpecific(
            String subNodePath,
            Class<N> dataType,
            FBFilterOption filterOption,
            ResultCallback<Updates<N>> onAdd,
            ResultCallback<String> onError
    ) {
        _listenTo(
                _getReferenceOfNode(subNodePath),
                false,
                dataType,
                filterOption,
                onAdd,
                onError
        );
    }

    private <T2> void _listenTo(
            DatabaseReference ref,
            boolean listenToAnyChange,
            Class<T2> dataType,
            FBFilterOption filterOption,
            ResultCallback<Updates<T2>> onUpdate,
            ResultCallback<String> onError
    ) {
        if (_listners == null) {
            _listners = new HashMap<>();
        }

        ChildEventListener listener;

        if (listenToAnyChange) {
            listener = new FirebaseDatabaseOp.ChildEventListener() {
                @Override
                void onAnyChange(DataSnapshot snapshot, Boolean wasAdded) {
                    onChangeReceived(dataType, snapshot, wasAdded, onUpdate, onError);
                }

                @Override
                public void onCancelled(@NonNull DatabaseError error) {
                    if (onError != null) {
                        onError.invoke(error.getMessage());
                    }
                }
            };
        }
        //
        else {
            listener = new FirebaseDatabaseOp.ChildEventListener() {
                @Override
                public void onChildAdded(@NonNull DataSnapshot snapshot, @androidx.annotation.Nullable String previousChildName) {
                    onChangeReceived(dataType, snapshot, true, onUpdate, onError);
                }

                @Override
                public void onCancelled(@NonNull DatabaseError error) {
                    if (onError != null) {
                        onError.invoke(error.getMessage());
                    }
                }
            };
        }

        if (filterOption == null) {
            _listners.put(getRefPath(ref), listener);
            ref.addChildEventListener(listener);
        }
        //
        else {
            Query query = buildQuery(ref, filterOption);
            _listners.put(getRefPath(ref) + "/" + filterOption.toString().hashCode(), listener);
            query.addChildEventListener(listener);
        }
    }

    private <T2> void onChangeReceived(
            Class<T2> dataType,
            DataSnapshot snapshot,
            Boolean wasAdded,
            ResultCallback<Updates<T2>> onUpdate,
            ResultCallback<String> onError
    ) {
        onChangeReceived(
                dataType,
                snapshot,
                wasAdded,
                null,
                onUpdate,
                onError
        );
    }

    private <T2> void onChangeReceived(
            Class<T2> dataType,
            DataSnapshot snapshot,
            Boolean wasAdded,
            ActionCallback<Object, Pair<T2, StringSet>> customConverter,
            ResultCallback<Updates<T2>> onUpdate,
            ResultCallback<String> onError
    ) {
        handleReceivedDataSnapshot(snapshot, dataType, customConverter, result -> {
            if (result.data != null) {
                onUpdate.invoke(new Updates<>(result.data, wasAdded));
            } else {
                if (onError != null) {
                    onError.invoke(
                            "Error in handling value: " + snapshot.getValue() +
                                    "\nKey: " + snapshot.getKey() +
                                    "\nDetails: " + snapshot.toString()
                    );
                }
            }
        });
    }

    //--------------------------------------------------------

    @Override
    public void removeListeners(String subNodePath, FBFilterOption filterOption) {
        var ref = _getReferenceOfNode(subNodePath);
        removeListeners(ref, filterOption);
    }

    public void removeListeners(DatabaseReference ref, FBFilterOption filterOption) {
        if (_listners == null) return;

        if (filterOption == null) {
            var listener = _listners.get(getRefPath(ref));
            if (listener == null) return;

            ref.removeEventListener(listener);
        } else {
            var listener = _listners.get(getRefPath(ref) + "/" + filterOption.toString().hashCode());
            if (listener == null) return;

            ref.removeEventListener(listener);
        }
    }

    //--------------------------------------------------------

    private Map<String, ChildEventListener> _listners;

    private static class ChildEventListener implements com.google.firebase.database.ChildEventListener {
        void onAnyChange(DataSnapshot snapshot, Boolean wasAdded) {
        }

        @Override
        public void onChildAdded(@NonNull DataSnapshot snapshot, @androidx.annotation.Nullable String previousChildName) {
            onAnyChange(snapshot, true);
        }

        @Override
        public void onChildRemoved(@NonNull DataSnapshot snapshot) {
            onAnyChange(snapshot, false);
        }

        @Override
        public void onChildChanged(@NonNull DataSnapshot snapshot, @androidx.annotation.Nullable String previousChildName) {
            onAnyChange(snapshot, null);
        }

        @Override
        public void onChildMoved(@NonNull DataSnapshot snapshot, @androidx.annotation.Nullable String previousChildName) {
            onAnyChange(snapshot, null);
        }

        @Override
        public void onCancelled(@NonNull DatabaseError error) {
        }
    }

    private String getRefPath(DatabaseReference ref) {
        StringBuilder sb = new StringBuilder();
        sb.append(ref.getKey());

        var parent = ref.getParent();
        while (parent != null) {
            sb.insert(0, "/");
            sb.insert(0, parent.getKey());
            parent = parent.getParent();
        }

        return sb.toString();
    }

    //----------------------------------------------------------------------------

    @Override
    public void clear(ResultCallback<Response<Boolean>> callback) {
        var ref = databaseReference;
        removeNode(ref, callback);
    }

    @Override
    public void removeNode(String subNodePath, ResultCallback<Response<Boolean>> callback) {
        var ref = _getReferenceOfNode(subNodePath);
        removeNode(ref, callback);
    }

    public void removeNode(DatabaseReference ref, ResultCallback<Response<Boolean>> callback) {
        ref.removeValue((error, ref1) -> {
            if (callback != null) {
                if (error == null) {
                    callback.invoke(Response.success(true));
                } else {
                    callback.invoke(Response.failed(new StringSet(error.getMessage())));
                }
            }
        });
    }

    //------------------------------------------------------------------------------

    private Query buildQuery(DatabaseReference ref, @NotNull FBFilterOption filterOption) {
        Query query = ref.orderByChild(filterOption.key);

        if (filterOption.type == FBFilterTypes.Equal) {
            if (filterOption.args instanceof String) {
                query = ref.equalTo((String) filterOption.args);
            }
            //
            else if (filterOption.args instanceof Double) {
                query = ref.equalTo((Double) filterOption.args);
            }
            //
            else if (filterOption.args instanceof Boolean) {
                query = ref.equalTo((Boolean) filterOption.args);
            }
        }
        //
        else if (filterOption.type == FBFilterTypes.GreaterThan) {
            if (filterOption.args instanceof String) {
                query = query.startAfter((String) filterOption.args);
            }
            //
            else if (filterOption.args instanceof Double) {
                query = query.startAfter((Double) filterOption.args);
            }
            //
            else if (filterOption.args instanceof Boolean) {
                query = query.startAfter((Boolean) filterOption.args);
            }
        }
        //
        else if (filterOption.type == FBFilterTypes.GreaterThanOrEqual) {
            if (filterOption.args instanceof String) {
                query = query.startAt((String) filterOption.args);
            }
            //
            else if (filterOption.args instanceof Double) {
                query = query.startAt((Double) filterOption.args);
            }
            //
            else if (filterOption.args instanceof Boolean) {
                query = query.startAt((Boolean) filterOption.args);
            }
        }
        //
        else if (filterOption.type == FBFilterTypes.LessThan) {
            if (filterOption.args instanceof String) {
                query = query.endBefore((String) filterOption.args);
            }
            //
            else if (filterOption.args instanceof Double) {
                query = query.endBefore((Double) filterOption.args);
            }
            //
            else if (filterOption.args instanceof Boolean) {
                query = query.endBefore((Boolean) filterOption.args);
            }
        }
        //
        else if (filterOption.type == FBFilterTypes.LessThanOrEqual) {
            if (filterOption.args instanceof String) {
                query = query.endAt((String) filterOption.args);
            }
            //
            else if (filterOption.args instanceof Double) {
                query = query.endAt((Double) filterOption.args);
            }
            //
            else if (filterOption.args instanceof Boolean) {
                query = query.endAt((Boolean) filterOption.args);
            }
        }

        if (filterOption.limit != null) {
            if (filterOption.limit.fromStart) {
                query = query.limitToFirst(filterOption.limit.count);
            } else {
                query = query.limitToLast(filterOption.limit.count);
            }
        }

        return query;
    }
}

File: GMUtilsFlutter/gmutils_flutter/lib/zgmutils/data_utils/firebase/firebase_database_op.dart
import 'dart:async';

import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_database/firebase_database.dart';
import 'package:ogtech_app_store/zgmutils/utils/text_utils.dart';

import '../../utils/logs.dart';
import '../../utils/mappable.dart';
import '../../utils/string_set.dart';
import 'firebase_utils.dart';
import 'response.dart';

abstract class IFirebaseDatabaseOp<T> {
  final Mappable<T> mappable;
  final String rootNodeName;

  IFirebaseDatabaseOp({
    required this.mappable,
    required this.rootNodeName,
  });

  //----------------------------------------------------------------------------

  Future<bool> isConnectionAvailable() {
    return FirebaseUtils.isConnectionAvailable();
  }

  //----------------------------------------------------------------------------

  Future<Response<bool>> saveData(T data, {required String? subNodePath});

  Future<Response<bool>> saveMultipleData({
    required Map<String, T> nodesAndData,
  });

  //----------------------------------------------------------------------------

  Future<Response<List<T>>> retrieveAll({
    FBFilterOption? filterOption,
    List<Map> Function(Object value)? collectionSource,
  });

  //----------------------------------------------------------------------------

  Future<Response<T>> retrieveOnly({
    required String? subNodePath,
  });

  //----------------------------------------------------------------------------

  void listenToChanges({
    required String? subNodePath,
    required Function(T) onChange,
    void Function()? onDone,
    void Function(Object, StackTrace)? onError,
  });

  void listenToChangesSpecific<N>({
    required String? subNodePath,
    required Function(N) onChange,
    void Function()? onDone,
    void Function(Object, StackTrace)? onError,
  });

  void listenToAdding({
    required String? subNodePath,
    required Function(T) onAdd,
    void Function()? onDone,
    void Function(Object, StackTrace)? onError,
  });

  void listenToAddingSpecific<N>({
    required String? subNodePath,
    required Function(N) onAdd,
    void Function()? onDone,
    void Function(Object, StackTrace)? onError,
  });

  void removeListeners({required String? subNodePath});

  //----------------------------------------------------------------------------

  Future<Response<bool>> clear();

  Future<Response<bool>> removeNode({required String subNodePath});
}

class FBFilterOption {
  final FBFilterTypes type;
  final String key;
  final Object? args;
  final int? limit;

  FBFilterOption({
    required this.type,
    required this.key,
    required this.args,
    required this.limit,
  });

  @override
  String toString() {
    return 'FBFilterOption{type: $type, key: $key, args: $args, limit: $limit}';
  }
}

enum FBFilterTypes {
  equal,
  greaterThan,
  greaterThanOrEqual,
  lessThan,
  lessThanOrEqual,
}

//++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

class FirebaseDatabaseOp<T> extends IFirebaseDatabaseOp<T> {
  FirebaseDatabaseOp({
    required super.mappable,
    required super.rootNodeName,
  });

  //----------------------------------------------------------------------------

  DatabaseReference? _databaseReference;

  Future<DatabaseReference> get databaseReference async {
    if (_databaseReference == null) {
      init() {
        var databaseReference = FirebaseDatabase.instance.ref();
        _databaseReference = databaseReference.child(rootNodeName);
      }

      try {
        init();
      } catch (e) {
        Logs.print(() => 'FirebaseDatabaseOp.databaseReference ---> Exception:: $e');
        await Firebase.initializeApp();
        init();
      }
    }

    return _databaseReference!;
  }

  //----------------------------------------------------------------------------

  String _refineKeyName(String name) => FirebaseUtils.refineKeyName(name);

  Future<DatabaseReference> _getReferenceOfNode(String? subNodePath) async {
    if (subNodePath == null) return databaseReference;

    subNodePath = FirebaseUtils.refinePathFragmentNames(subNodePath);

    var ref = await databaseReference;
    return ref.child(subNodePath);
  }

  //----------------------------------------------------------------------------

  @override
  Future<Response<bool>> saveData(T data, {required String? subNodePath}) async {
    DatabaseReference ref = await _getReferenceOfNode(subNodePath);
    return saveDataTo(ref, data);
  }

  Future<Response<bool>> saveDataTo(DatabaseReference ref, T data) async {
    try {
      bool added = false;

      await ref
          .set(mappable.toMap(data))
          .whenComplete(() async => added = true)
          .onError((error, stackTrace) async => added = false);

      Logs.print(() => 'FirebaseDatabaseOp.saveDataTo'
          '(ref: ${ref.path}, data: ${TextUtils().trimEnd('$data')}) '
          '---> added: $added');

      return Response.success(data: added);
    } catch (e) {
      Logs.print(() => 'FirebaseDatabaseOp.saveDataTo(ref: ${ref.path}) ---> Exception:: $e');
      return Response.failed(
        error: StringSet(e.toString()),
        connectionFailed: true,
      );
    }
  }

  @override
  Future<Response<bool>> saveMultipleData({required Map<String, T> nodesAndData,}) async {
    DatabaseReference ref = await _getReferenceOfNode(null);
    return saveMultipleDataTo(ref, nodesAndData: nodesAndData);
  }

  Future<Response<bool>> saveMultipleDataTo(
      DatabaseReference ref, {
        required Map<String, T> nodesAndData,
      }) async {
    try {
      bool added = false;

      var nodesAndMappedData = nodesAndData.map(
            (key, value) => MapEntry(
          _refineKeyName(key),
          mappable.toMap(value),
        ),
      );

      await ref
          .set(nodesAndMappedData)
          .whenComplete(() async => added = true)
          .onError((error, stackTrace) async => added = false);

      Logs.print(() => 'FirebaseDatabaseOp.saveMultipleDataTo'
          '(ref: ${ref.path}, data-length: ${nodesAndData.length}, data: ${TextUtils().trimEnd('$nodesAndData')}) '
          '---> added: $added');

      return Response.success(data: added);
    } catch (e) {
      Logs.print(() => 'FirebaseDatabaseOp.saveMultipleDataTo(ref: ${ref.path}) ---> Exception:: $e');
      return Response.failed(
        error: StringSet(e.toString()),
        connectionFailed: true,
      );
    }
  }

  //----------------------------------------------------------------------------

  @override
  Future<Response<List<T>>> retrieveAll({
    FBFilterOption? filterOption,
    List<Map> Function(Object value)? collectionSource,
  }) async {
    if (await isConnectionAvailable() == false) {
      return Response.failed(
        error: StringSet('No Connection', 'لا يوجد اتصال'),
        connectionFailed: true,
      );
    }

    var ref = await databaseReference;
    DataSnapshot? snapshot;

    try {
      Query? query;
      if (filterOption != null) {
        query = ref.orderByChild(filterOption.key);

        if (filterOption.type == FBFilterTypes.equal) {
          query = query.equalTo(filterOption.args);
        }
        //
        else if (filterOption.type == FBFilterTypes.greaterThan) {
          query = query.startAfter(filterOption.args);
        }
        //
        else if (filterOption.type == FBFilterTypes.greaterThanOrEqual) {
          query = query.startAt(filterOption.args);
        }
        //
        else if (filterOption.type == FBFilterTypes.lessThan) {
          query = query.endBefore(filterOption.args);
        }
        //
        else if (filterOption.type == FBFilterTypes.lessThanOrEqual) {
          query = query.endAt(filterOption.args);
        }

        if (filterOption.limit != null) {
          query = query.limitToFirst(filterOption.limit!);
        }
      }

      if (query == null) {
        var result = await ref.once();
        snapshot = result.snapshot;
      }
      //
      else {
        //snapshot = await query.get();
        var result = await query.once();
        snapshot = result.snapshot;
      }


      Response<List<T>> response;
      
      if (snapshot.exists) {
        List<T> list = await _mapData(
          snapshot,
          collectionSource,
        );

        response = Response.success(data: list);
      }
      //
      else {
        response = Response.failed(
          error: StringSet('No data', "لا توجد بيانات"),
        );
      }

      Logs.print(() => 'FirebaseDatabaseOp.retrieveAll'
          '(ref: ${ref.path}, filterOption: $filterOption) '
          '---> '
          'response.data-length: ${response.data?.length}, '
          'response.data: ${TextUtils().trimEnd('${response.data}')}, '
          'response.message: ${response.error}');
      
      return response;
    } catch (e) {
      Logs.print(() => 'FirebaseDatabaseOp.retrieveAll(ref: ${ref.path}) ---> snapshot: ${snapshot?.value} ---> Exception:: $e');
      return Response.failed(
        error: StringSet(e.toString()),
        connectionFailed: true,
      );
    }
  }

  Future<List<T>> _mapData(
      DataSnapshot dataSnapshot,
      List<Map> Function(Object value)? collectionSource,
      ) async {
    List<T> list = [];

    if (dataSnapshot.value != null) {
      dynamic values;

      if (collectionSource == null) {
        if (dataSnapshot.value is List) {
          values = dataSnapshot.value;
        }
        //
        else {
          List<Map> list = [];
          for (var entry in (dataSnapshot.value as Map).entries) {
            list.add(entry.value);
          }
          values = list;
        }
      }
      //
      else {
        values = collectionSource(dataSnapshot);
      }

      for (dynamic item in values) {
        var map = Map<String, dynamic>.from(item);
        list.add(mappable.fromMap(map));
      }
    }

    return list;
  }

  //----------------------------------------------------------------------------

  @override
  Future<Response<T>> retrieveOnly({required String? subNodePath}) async {
    var ref = await _getReferenceOfNode(subNodePath);
    return await retrieveOnlyFrom(ref: ref);
  }

  Future<Response<T>> retrieveOnlyFrom({
    required DatabaseReference ref,
  }) async {
    if (await isConnectionAvailable() == false) {
      return Response.failed(
        error: StringSet('No Connection', 'لا يوجد اتصال'),
        connectionFailed: true,
      );
    }

    var map;

    try {
      var event = await ref.once();
      var snapshot = event.snapshot;

      Response<T> response;
      
      if (snapshot.exists) {
        if (snapshot.value is List) {
          var data = snapshot.value as List;
          map = Map<String, dynamic>.from(data[0]);
        }
        //
        else {
          var data = snapshot.value as Map;
          var nodeName = ref.path.substring(ref.path.lastIndexOf("/") + 1);
          if (data[nodeName] == null) {
            map = Map<String, dynamic>.from(data);
          } else {
            map = Map<String, dynamic>.from(data[nodeName]);
          }
        }

        response = Response.success(data: mappable.fromMap(map));
      }
      //
      else {
        response = Response.failed(
          error: StringSet('No data', "لا توجد بيانات"),
        );
      }

      Logs.print(() => 'FirebaseDatabaseOp.retrieveOnlyFrom(ref: ${ref.path}) '
          '---> '
          'response.data: ${TextUtils().trimEnd('${response.data}')}, '
          'response.message: ${response.error}',
      );

      return response;
    } catch (e) {
      Logs.print(() => 'FirebaseDatabaseOp.retrieveOnlyFrom(ref: ${ref.path}) ---> result: $map ---> Exception:: $e');
      
      return Response.failed(
        error: StringSet(e.toString()),
        connectionFailed: true,
      );
    }
  }

  //----------------------------------------------------------------------------

  @override
  void listenToChanges({
    required String? subNodePath,
    required Function(T) onChange,
    void Function()? onDone,
    void Function(Object, StackTrace)? onError,
  }) async {
    return listenToChangesOn(
      ref: await _getReferenceOfNode(subNodePath),
      onChange: onChange,
      onDone: onDone,
      onError: onError,
    );
  }

  void listenToChangesOn({
    required DatabaseReference ref,
    required Function(T) onChange,
    void Function()? onDone,
    void Function(Object, StackTrace)? onError,
  }) async {
    return _listenTo(
      ref: ref,
      listenToAnyChange: true,
      onNewUpdate: onChange,
      onDone: onDone,
      onError: onError,
    );
  }

  @override
  void listenToAdding({
    required String? subNodePath,
    required Function(T) onAdd,
    void Function()? onDone,
    void Function(Object, StackTrace)? onError,
  }) async {
    return listenToAddingTo(
        ref: await _getReferenceOfNode(subNodePath),
        onAdd: onAdd,
        onDone: onDone,
        onError: onError);
  }

  void listenToAddingTo({
    required DatabaseReference ref,
    required Function(T) onAdd,
    void Function()? onDone,
    void Function(Object, StackTrace)? onError,
  }) async {
    return _listenTo(
      ref: ref,
      listenToAnyChange: false,
      onNewUpdate: onAdd,
      onDone: onDone,
      onError: onError,
    );
  }

  void _listenTo({
    required DatabaseReference ref,
    required bool listenToAnyChange,
    required Function(T) onNewUpdate,
    void Function()? onDone,
    void Function(Object, StackTrace)? onError,
  }) async {
    onData(event) => (event) {
      Map<String, dynamic>? map;
      try {
        var snapshot = event.snapshot;
        map = Map.from(snapshot.value as Map);
        onNewUpdate(mappable.fromMap(map));
      } catch (e) {
        Logs.print(
              () => '***** FirebaseDatabaseOp.setOnChildAddedListener() **** '
              'args(into: ${ref.path}) \n'
              'result: $map \n'
              '$e',
        );
      }
    };

    if (listenToAnyChange) {
      ref.onChildChanged.listen(onData, onDone: onDone, onError: onError);
    } else {
      ref.onChildAdded.listen(onData, onDone: onDone, onError: onError);
    }
  }

  @override
  void removeListeners({required String? subNodePath}) async {
    var ref = await _getReferenceOfNode(subNodePath);
    removeListenersOf(ref: ref);
  }

  void removeListenersOf({required DatabaseReference ref}) {
    ref.onChildChanged.listen(null);
    ref.onChildAdded.listen(null);
  }

  //----------------------------------------------------------------------------

  @override
  Future<Response<bool>> clear() async {
    try {
      bool suc = true;
      var ref = await databaseReference;
      await ref.remove().onError((error, stackTrace) => suc = false);
      Logs.print(() => 'FirebaseDatabaseOp.clear() ---> success:: $suc');
      return Response.success(data: suc);
    } catch (e) {
      Logs.print(() => 'FirebaseDatabaseOp.clear() ---> Exception:: $e');
      return Response.failed(
        error: StringSet(e.toString()),
        connectionFailed: true,
      );
    }
  }

  @override
  Future<Response<bool>> removeNode({required String subNodePath}) async {
    var ref = await _getReferenceOfNode(subNodePath);
    try {
      bool deleted = true;
      await ref.remove().onError((error, stackTrace) => deleted = false);
      Logs.print(() => 'FirebaseDatabaseOp.removeNode(ref: ${ref.path}) ---> deleted:: $deleted');
      return Response.success(data: deleted);
    } catch (e) {
      Logs.print(() => 'FirebaseDatabaseOp.removeNode(ref: ${ref.path}) ---> Exception:: $e');
      return Response.failed(
        error: StringSet(e.toString()),
        connectionFailed: true,
      );
    }
  }

  @override
  void listenToAddingSpecific<N>({required String? subNodePath, required Function(N p1) onAdd, void Function()? onDone, void Function(Object p1, StackTrace p2)? onError}) {
    throw 'implement listenToAddingSpecific';
  }

  @override
  void listenToChangesSpecific<N>({required String? subNodePath, required Function(N p1) onChange, void Function()? onDone, void Function(Object p1, StackTrace p2)? onError}) {
    throw 'implement listenToChangesSpecific';
  }
}
